addons/website_slides/static/src/js/slides_course_page.js

```diff
@@ -0,0 +1,148 @@
+     */
+    toggleCompletionButton: function (slide, completed = true) {
+        const $button = this.$(`.o_wslides_sidebar_done_button[data-id=${slide.id}]`);
+
+        const newButton = QWeb.render('website.slides.sidebar.done.button', {
+            slideId: slide.id,
+            slideCompleted: completed,
+            canSelfMarkUncompleted: slide.canSelfMarkUncompleted,
+            canSelfMarkCompleted: slide.canSelfMarkCompleted,
+            isMember: slide.isMember,
+        });
+
+        if ($button.length) {
+            $button.replaceWith(newButton);
+            const $elem = this.$(`.o_wslides_sidebar_done_button[data-id="${slide.id}"]`);
+            $elem.data('completed', completed);
+        }
+    },
+
+    /**
+     * Updates the progressbar whenever a lesson is completed
+     *
+     * @public
+     * @param {*} channelCompletion
+     */
+    updateProgressbar: function (channelCompletion) {
+        const completion = Math.min(100, channelCompletion);
+
+        const $completed = $('.o_wslides_channel_completion_completed');
+        const $progressbar = $('.o_wslides_channel_completion_progressbar');
+
+        if (completion < 100) {
+            // Hide the "Completed" text and show the progress bar
+            $completed.addClass('d-none');
+            $progressbar.removeClass('d-none').addClass('d-flex');
+        } else {
+            // Hide the progress bar and show the "Completed" text
+            $completed.removeClass('d-none');
+            $progressbar.addClass('d-none').removeClass('d-flex');
+        }
+
+        $progressbar.find('.progress-bar').css('width', `${completion}%`);
+        $progressbar.find('.o_wslides_progress_percentage').text(completion);
+    },
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    /**
+     * Once the completion conditions are filled,
+     * rpc call to set the relation between the slide and the user as "completed"
+     *
+     * @private
+     * @param {Integer} slideId: the id of slide to set as completed
+     * @param {Boolean} completed: true to mark the slide as completed
+     *     false to mark the slide as not completed
+     */
+    _toggleSlideCompleted: async function (slide, completed = true) {
+        if (!!slide.completed === !!completed || !slide.isMember) {
+            // no useless RPC call
+            return;
+        }
+
+        const data = await this._rpc({
+            route: `/slides/slide/${completed ? 'set_completed' : 'set_uncompleted'}`,
+            params: {slide_id: slide.id},
+        });
+
+        this.toggleCompletionButton(slide, completed);
+        this.updateProgressbar(data.channel_completion);
+
+        slide.completed = true;
+    },
+
+    _getSlide: function (slideId) {
+        return $(`.o_wslides_sidebar_done_button[data-id="${slideId}"]`).data();
+    },
```

Comment:
Seems strange that this uses DOM information.

Somehow wondering if we should not delegate finding the slide data based on the id to the Widgets extending this one.

We could have an empty "_getSlideData" that needs to be overridden, that way you can override is and use `this.get('slide')` I'm guessing, which will look into the Widget data.
(And/or some other means for other extending Widgets that maybe don't have a `this.get('slide')` available).

Reply:
Well the update of the progress bar is less critical, it's simply UI stuff, if it's not there jQuery will not do anything.

Then maybe we can find a common ground, and override the "_getSlideData" method to use `this.get('slide')` when possible in the related sub-widget (I think it's the fullscreen one)?